### DISCUSSION TITLE
Automatically scan all ECU's for DTCs

### DISCUSSION THREAD

**Discussion by emcodem** [cedricp/ddt4all#852] [Q&A]
After selling my old VW and buying a used Dacia, i was very pleased to see that there is this nice project on Github, thank you so much for it! Is this all reverse engineered or did renault also support this (did you ever ask for support from them?, obviously they accept that it exists)?

Anyway, so today i used it first time and wanted to find out why some warning light is on. I learned that i had to select manually each and every ecu (and submodules?) and hit the scan button, then click accept the "no dtc found" alert box and select next ECU. 

So i wanted to ask: is there a special reason why there is no "automatic dtc scan over all modules" (other than its not yet implemented)? I guess it is a pretty trivial thing to do but maybe i am totally wrong, i didnt yet look very deep into the source code...

**Comment by Furtif** [comment]
this is not a issue .. convert to discussion

**Comment by Furtif** [comment]
done

**Comment by Furtif** [comment]
https://github.com/cedricp/ddt4all#informations

**Reply by emcodem** [reply]
Thanks, i should have seen that this repo uses discussions in the readme.md... but also when re-reading i don't find an answer for the auto dtc scan question... You think i overlooked it?

**Reply by Furtif** [reply]
I leave this answer to the community.. as far as I know... I know as much as you

**Comment by KarelSvo** [comment]
> and buying a used Dacia,

Since you didn't exactly define the Dacia.
Vehicles before 2013 mostly have fail flag control units. The error definition is located there in the on-screen menus.
Pyren or pyclip are better suited for error diagnosis. https://gitlab.com/py_ren/pyren     
https://github.com/Marianpol/pyclip
You also do not specify the type of warning light. Vehicles from 2013 and later activate the electronic warning light when the service interval is exceeded.

**Reply by emcodem** [reply]
Hi Karel, thanks for your reply!

No worries about my light, eventually i found the reason for it (oil service overdue), my question is more theoretical and developer nature because  most forum entries just told to select each and every ecu manually and scan - and i though that it should be a trivial task to re-use ddt4all/pyren code and write "scan all" script.
I tried Pyren Rescan option but it was so fast that i did not trust that it collected DTC's from all modules.
But what you say is that all DTC's are collected by the car already? ..So in my case, Logan 2, i would normally find all DTC's in Cluster/TDB?

**Reply by shrlnm** [reply]
When pyren rescans DTCs it just get the number of DTCs in every ECUs and it really don't take too much time. But it's only count of DTC. Before showing DTCs to user we have to load full ECU data from DB for filtering uninteresting DTC and for reading DTC context and their freeze frame. It may take significant time.

**Reply by emcodem** [reply]
@shrlnm Thanks a lot for this info, it is very interesting that we can get the count of DTC very quick obviously. @KarelSvo you mention "vehicles before 2013 have fail flag control units, what about cars after? From Feeling my 2017 Logan2 seems to collect all DTC's in Cluster/TBD section...

**Reply by shrlnm** [reply]
@emcodem I don't know if failflag ecus are installed on recent cars or not. But pyren marks such ECUs on a main screen with tag "failflag" or show 'n/a' after DTCs rescan. Regarding collection all DTCs in a TDB  it's  ofcourse  misunderstanding. TDB do not collect DTCs from other ECUs. There are special system CAN frames which allows to some ECUs to ask TDB to fire MIL lamp.

**Reply by emcodem** [reply]
That is extremely good info, thanks a bunch!

**Comment by cedricp** [comment]
Short answer : ddt4all is more an engineering tool (to study ECUs), whereas PyRen is a (very good) diagnostic tool. Keep in mind that I started this project "just" to repair my Megane 2 airbags fault :)

**Reply by emcodem** [reply]
Hehe you invested lots of time to repair your airbags fault :D 
Honestly i'd be interested in the "long answer" but i don't want to steal too much of your time, just curious... You know, your tool is well received out there and it could also have been that you say something like "patch welcome" (in that case i'd probably look into it) ...again just curious :D

**Comment by cedricp** [comment]
There's no long story ! You remember Forrest gump who went for a little run and finally figured since he'd gone this far, then he kept going ? That's approximately the same for me, LOL ! More seriously, I had the feeling that I was doing something that could be appreciated by others.

**Reply by emcodem** [reply]
Hehe what a nice comparison :D From what i read so far nothing really speaks against automated DTC scan but reading what Karel says above makes me doubt that i can do it... i have no problem with python coding but i don't know anything at all about the cars side

**Reply by cedricp** [reply]
Yes, sorry, it's possible to do it once all the ECUs have been scanned. I'll look into that when I have a moment.

**Comment by KarelSvo** [comment]
The Logan 2 has no fail flag control units.
ddtall_py3 does not show any scan results in the log window when scanning the KWP ecus.
The Autoident data is only in the log file.
When scanning, ddt4all_py3 skips the 29 bit addresses.
ddt4all_py3 has problems with mapping model code larger than 3 digits. For example x79Ph2.

**Reply by cedricp** [reply]
Do you mean KWP and 29bits CAN are no longer detected since py3 ?
Can you tell me more about that ?

**Reply by Furtif** [reply]
I thinks uses code py2.7 new code get Car by split(“-“)[0] and old uses var[:-3] more or less

**Reply by Furtif** [reply]
https://github.com/cedricp/ddt4all/blob/master/main.py#L161

**Reply by Furtif** [reply]
All old Car list before full update pr is code Car as len 3 - name new list as some as len up of 3 and code is set to get as split- and not init string.

**Comment by KarelSvo** [comment]
https://github.com/cedricp/ddt4all/blob/master/ecu.py#L1619  
Skipping 29 bit addresses was also in the py2 version.

py3        https://github.com/cedricp/ddt4all/blob/master/ecu.py#L1702
The result of check_ecu2 is only displayed with the CAN scan, not with the KWP scan. In the py2 version no problem.
The KWP identification 2180 data are in the ecu_ddt protocol.

**Reply by cedricp** [reply]
Yes, 29bits is not supported yet, indeed. But I can't see any problem with KWP, and I don't have a car with that protocol to test.

**Reply by cedricp** [reply]
About KWP, could you try to uncomment line 1340 and test :
`# self.cmd("AT WS")`
It has been commented for unknown reason...

**Reply by Furtif** [reply]
i not have touch in @cedricp  :|

**Reply by Furtif** [reply]
if ok fix .... or user pr it

**Reply by Furtif** [reply]
For other side i have test OBDLink EX seen ok in test button
if results ok need edit [Readme](https://github.com/cedricp/ddt4all/blob/master/README.md#supported-diagnostic-adapters-so-far) 
set OBDlink SX/EX

```
v1.0 ATZ [OK]
v1.0 ATE1 [OK]
v1.0 ATL0 [OK]
v1.0 ATI [OK]
v1.0 AT@1 [OK]
v1.0 ATAL [OK]
v1.0 ATBD [OK]
v1.0 ATBI [OK]
v1.0 ATCAF0 [OK]
v1.0 ATCFC1 [OK]
v1.0 ATCP 01 [OK]
v1.0 ATCS [OK]
v1.0 ATCV 1250 [OK]
v1.0 ATD [OK]
v1.0 ATDP [OK]
v1.0 ATDPN [OK]
v1.0 ATH0 [OK]
v1.0 ATI [OK]
v1.0 ATIB 10 [OK]
v1.0 ATIB 96 [OK]
v1.0 ATL1 [OK]
v1.0 ATM0 [OK]
v1.0 ATMA [OK/TEMPS DEPASSÉ]
v1.0 ATMR 01 [OK/TEMPS DEPASSÉ]
v1.0 ATMT 01 [OK/TEMPS DEPASSÉ]
v1.0 ATNL [OK/TEMPS DEPASSÉ]
v1.0 ATPC [OK]
v1.0 ATR1 [OK]
v1.0 ATRV [OK]
v1.0 ATSH 001122 [OK]
v1.0 ATSH 012 [OK]
v1.0 ATSP A6 [OK]
v1.0 ATSP 6 [OK]
v1.0 ATCM 123 [OK]
v1.0 ATCM 12345678 [OK]
v1.0 ATCF 123 [OK]
v1.0 ATCF 12345678 [OK]
v1.0 ATST FF [OK]
v1.0 ATSW 96 [OK]
v1.0 ATSW 34 [OK]
v1.0 ATTP A6 [OK]
v1.0 ATTP 6 [OK]
v1.0 ATWM 817AF13E [OK]
v1.0 ATWS [OK]
v1.1 ATFC SD 300000 [OK]
v1.1 ATFC SH 012 [OK]
v1.1 ATFC SH 00112233 [OK]
v1.1 ATFC SM 1 [OK]
v1.1 ATPP FF OFF [OK]
v1.1 ATPP FF ON [OK]
v1.1 ATPPS [OK]
v1.2 ATAR [OK]
v1.2 ATAT1 [OK]
v1.2 ATDM1 [OK/TEMPS DEPASSÉ]
v1.2 ATIFR H [OK/TEMPS DEPASSÉ]
v1.2 ATIFR0 [OK]
v1.2 ATIIA 01 [OK]
v1.2 ATKW0 [OK]
v1.2 ATMP 0123 [OK/TEMPS DEPASSÉ]
v1.2 ATMP 0123 4 [OK/TEMPS DEPASSÉ]
v1.2 ATMP 012345 [OK/TEMPS DEPASSÉ]
v1.2 ATMP 012345 6 [OK/TEMPS DEPASSÉ]
v1.2 ATSR 01 [OK]
v1.3 ATCRA 012 [OK]
v1.3 ATCRA 01234567 [OK]
v1.3 ATD0 [OK]
v1.3 ATFE [OK]
v1.3 ATJE [OK]
v1.3 ATJS [OK]
v1.3 ATKW [OK]
v1.3 ATRA 01 [OK]
v1.3 ATS1 [OK]
v1.3 ATSP 00 [OK]
v1.3 ATV0 [OK]
v1.4 ATCEA [OK]
v1.4 ATCEA 01 [OK]
v1.4 ATCV 0000 [OK]
v1.4 ATIB 48 [OK]
v1.4 ATIGN [OK]
v1.4 ATPB 01 23 [OK]
v1.4 ATRD [OK]
v1.4 ATSD 01 [OK]
v1.4 ATSI [OK]
v1.4 ATFI [OK]
v1.4 ATSS [OK]
v1.4 ATTA 12 [OK]
v1.4 ATCSM1 [OK]
v1.4 ATJHF1 [OK]
v1.4 ATJTM1 [OK]
v1.4 ATJTM5 [OK]
v1.4b ATCRA [OK]
v2.0 ATAMC [ECHEC]
v2.0 ATAMT 20 [ECHEC]
v2.1 ATCTM1 [ECHEC]
v2.1 ATCTM5 [ECHEC]
Résultat: 91 réussis sur 95
Version ELM maximum:v1.4b
```
